File: libs/install/src/adapters.ts
import { join } from 'node:path';

export type Scope = 'project' | 'global';

/** Base directories each scope resolves against. */
export interface Bases {
  /** The current project root (scope `project`). */
  project: string;
  /** The user's home directory (scope `global`). */
  home: string;
}

interface SkillAdapter {
  scopes: Scope[];
  dest: (scope: Scope, id: string, bases: Bases) => string;
}

/** Agents that can receive a skill, and where the skill lands. */
const SKILL_ADAPTERS: Record<string, SkillAdapter> = {
  claude: {
    scopes: ['project', 'global'],
    dest: (scope, id, b) => join(scope === 'project' ? b.project : b.home, '.claude', 'skills', id),
  },
  codex: {
    scopes: ['project', 'global'],
    dest: (scope, id, b) => join(scope === 'project' ? b.project : b.home, '.agents', 'skills', id),
  },
  copilot: {
    scopes: ['project', 'global'],
    dest: (scope, id, b) =>
      scope === 'project'
        ? join(b.project, '.github', 'skills', id)
        : join(b.home, '.copilot', 'skills', id),
  },
  cursor: {
    scopes: ['project'],
    dest: (_scope, id, b) => join(b.project, '.cursor', 'rules', `${id}.mdc`),
  },
};

/** Agents whose skill install is a single `.mdc` file rather than a folder copy. */
export const SINGLE_FILE_AGENTS = new Set(['cursor']);

/** Resolve the on-disk destination for a skill, or throw if unsupported. */
export function resolveSkillDestination(
  agent: string,
  scope: Scope,
  id: string,
  bases: Bases,
): string {
  const adapter = SKILL_ADAPTERS[agent];
  if (!adapter) {
    throw new Error(`Agent "${agent}" does not support skills`);
  }
  if (!adapter.scopes.includes(scope)) {
    throw new Error(`Agent "${agent}" supports ${adapter.scopes.join(', ')} scope only`);
  }
  return adapter.dest(scope, id, bases);
}


export type PromptFormat = 'md' | 'copilot' | 'toml';

interface PromptAdapter {
  scopes: Scope[];
  format: PromptFormat;
  dest: (scope: Scope, id: string, bases: Bases) => string;
}

/** Agents that can receive a prompt, where it lands, and the format to render. */
const PROMPT_ADAPTERS: Record<string, PromptAdapter> = {
  claude: {
    scopes: ['project', 'global'],
    format: 'md',
    dest: (scope, id, b) =>
      join(scope === 'project' ? b.project : b.home, '.claude', 'commands', `${id}.md`),
  },
  codex: {
    scopes: ['global'],
    format: 'md',
    dest: (_scope, id, b) => join(b.home, '.codex', 'prompts', `${id}.md`),
  },
  cursor: {
    scopes: ['project', 'global'],
    format: 'md',
    dest: (scope, id, b) =>
      join(scope === 'project' ? b.project : b.home, '.cursor', 'commands', `${id}.md`),
  },
  copilot: {
    scopes: ['project', 'global'],
    format: 'copilot',
    dest: (scope, id, b) =>
      scope === 'project'
        ? join(b.project, '.github', 'prompts', `${id}.prompt.md`)
        : join(b.home, '.copilot', 'prompts', `${id}.prompt.md`),
  },
  gemini: {
    scopes: ['project', 'global'],
    format: 'toml',
    dest: (scope, id, b) =>
      join(scope === 'project' ? b.project : b.home, '.gemini', 'commands', `${id}.toml`),
  },
};

function promptAdapterFor(agent: string): PromptAdapter {
  const adapter = PROMPT_ADAPTERS[agent];
  if (!adapter) {
    throw new Error(`Agent "${agent}" does not support prompts`);
  }
  return adapter;
}

/** Resolve the on-disk destination for a prompt, or throw if unsupported. */
export function resolvePromptDestination(
  agent: string,
  scope: Scope,
  id: string,
  bases: Bases,
): string {
  const adapter = promptAdapterFor(agent);
  if (!adapter.scopes.includes(scope)) {
    throw new Error(`Agent "${agent}" supports ${adapter.scopes.join(', ')} scope only for prompts`);
  }
  return adapter.dest(scope, id, bases);
}

/** The render format for an agent's prompts. */
export function promptFormat(agent: string): PromptFormat {
  return promptAdapterFor(agent).format;
}

File: apps/cli-dotnet/AiSkills.Cli/Core/Adapters.cs
namespace AiSkills.Cli.Core;

public enum Scope
{
    Project,
    Global,
}

public record Bases(string Project, string Home);

public static class Adapters
{
    private static readonly HashSet<string> SingleFileAgents = new() { "cursor" };

    public static bool IsSingleFile(string agent) => SingleFileAgents.Contains(agent);

    /// <summary>Resolve the on-disk destination for a skill, or throw if unsupported.</summary>
    public static string ResolveSkillDestination(string agent, Scope scope, string id, Bases bases)
    {
        var baseDir = scope == Scope.Project ? bases.Project : bases.Home;

        return agent switch
        {
            "claude" => Path.Combine(baseDir, ".claude", "skills", id),
            "codex" => Path.Combine(baseDir, ".agents", "skills", id),
            "copilot" => scope == Scope.Project
                ? Path.Combine(bases.Project, ".github", "skills", id)
                : Path.Combine(bases.Home, ".copilot", "skills", id),
            "cursor" => scope == Scope.Project
                ? Path.Combine(bases.Project, ".cursor", "rules", $"{id}.mdc")
                : throw new ArgumentException("Agent \"cursor\" supports project scope only"),
            _ => throw new ArgumentException($"Agent \"{agent}\" does not support skills"),
        };
    }


    /// <summary>Resolve the destination for a prompt, or throw if unsupported.</summary>
    public static string ResolvePromptDestination(string agent, Scope scope, string id, Bases bases)
    {
        var baseDir = scope == Scope.Project ? bases.Project : bases.Home;

        return agent switch
        {
            "claude" => Path.Combine(baseDir, ".claude", "commands", $"{id}.md"),
            "codex" => scope == Scope.Global
                ? Path.Combine(bases.Home, ".codex", "prompts", $"{id}.md")
                : throw new ArgumentException("Agent \"codex\" supports global scope only for prompts"),
            "cursor" => Path.Combine(baseDir, ".cursor", "commands", $"{id}.md"),
            "copilot" => scope == Scope.Project
                ? Path.Combine(bases.Project, ".github", "prompts", $"{id}.prompt.md")
                : Path.Combine(bases.Home, ".copilot", "prompts", $"{id}.prompt.md"),
            "gemini" => Path.Combine(baseDir, ".gemini", "commands", $"{id}.toml"),
            _ => throw new ArgumentException($"Agent \"{agent}\" does not support prompts"),
        };
    }

    /// <summary>The render format for an agent's prompts: "md", "copilot", or "toml".</summary>
    public static string PromptFormat(string agent) => agent switch
    {
        "claude" or "codex" or "cursor" => "md",
        "copilot" => "copilot",
        "gemini" => "toml",
        _ => throw new ArgumentException($"Agent \"{agent}\" does not support prompts"),
    };
}
